dask/order.py

```diff
@@ -202,125 +193,212 @@ def dependencies_key(x):
-    outer_stack_seeds_append = outer_stack_seeds.append
-    inner_stack_append = inner_stack.append
-    inner_stack_pop = inner_stack.pop
-    inner_stack_extend = inner_stack.extend
+    outer_stack_pop = outer_stack.pop
+
+    next_nodes = {}
+    later_nodes = {}
+
+    # aliases for speed
     set_difference = set.difference
 
     is_init_sorted = False
-    item = min(init_stack, key=initial_stack_key)
     while True:
-        outer_stack_append(item)
-        while outer_stack:
-            item = outer_stack_pop()
-            if item not in result:
-                inner_stack_append(item)
-
-                while inner_stack:
-                    # Perform a DFS along dependencies until we complete our tactical goal
-                    item = inner_stack_pop()
-                    if item in result:
-                        continue
-
-                    if num_needed[item]:
-                        inner_stack_append(item)
-                        deps = set_difference(dependencies[item], result)
-                        if 1 < len(deps) < 1000:
-                            inner_stack_extend(
-                                sorted(deps, key=dependencies_key, reverse=True)
-                            )
+        while inner_stacks:
+            inner_stack = inner_stacks_pop()
+            while inner_stack:
+                # Perform a DFS along dependencies until we complete our tactical goal
+                item = inner_stack.pop()
+                if item in result:
+                    continue
+                if num_needed[item]:
+                    inner_stack.append(item)
+                    deps = set_difference(dependencies[item], result)
+                    if 1 < len(deps) < 1000:
+                        inner_stack.extend(
+                            sorted(deps, key=dependencies_key, reverse=True)
+                        )
+                    else:
+                        inner_stack.extend(deps)
+                    continue
+
+                result[item] = i
+                i += 1
+                deps = dependents[item]
+
+                if metrics[item][3] == 2:  # min_height
+                    # Don't leave any dangling single nodes!  Finish all dependents that are
+                    # ready and are also root nodes.  Doing this here also lets us continue
+                    # down to a different root node.
+                    finish_now = {
+                        dep
+                        for dep in deps
+                        if not dependents[dep] and num_needed[dep] == 1
+                    }
+                    if finish_now:
+                        deps -= finish_now  # Safe to mutate
+                        if len(finish_now) > 1:
+                            finish_now = sorted(finish_now, key=finish_now_key)
+                        for dep in finish_now:
+                            result[dep] = i
+                            i += 1
```

Comment:
I notice that this block isn't hit by our existing test. I also didn't see it in your notebook.

Is it not needed anymore? Or (random guess) perhaps your recent change to graph_metrics in https://github.com/dask/dask/pull/5872/commits/f5c5fec0dad9a8d3e6e9992b2065bacada5ebfdf#diff-d26fc8ea46375896e990e8f20828457fL537 introduced an off by one error in the creation of finish_now above?

Reply:
Aha, yes, off by one error.  Good catch!